### ISSUE TITLE
Doubly-linked home files break systemd unit names

### ISSUE THREAD

**Issue by nonsequitur** [rycee/home-manager#123]
For the `home-manager`-generated unit definitions that consist of two symlinks, systemd uses the name of the linked file (the store file name) as the unit name instead of the
name of the symlink.

To reproduce:
```bash
cat << 'EOF' > /tmp/showBug
{ pkgs, ... }:
{
  systemd.user.services.myservice = {
    Service.ExecStart = ''${pkgs.coreutils}/bin/true'';
    Install.WantedBy = [ "default.target" ];
  };
}
EOF
home-manager -f /tmp/showBug switch

systemctl --user daemon-reload

systemctl --user --all | grep myservice # => h7r8jgwzwvscv731852md3wxnwxv65aw-myservice.service
```

Sorry for not catching this earlier, I focused mainly on the single-link proposal that I deemed most promising.

Possible solutions:
- Add an option to copy files (very ugly)
- Find a way to fix the systemd behavior
- Use my original single-link proposal, maybe with the home file name included in the store name
  to address your shell globbing issues. I'll code it up quickly if you like.

**Comment by rycee** [comment]
Good catch. Thanks! For the shell globbing issue, I'm not sure how the store name could be manipulated in the case where `source` is set directly. I.e., for `home.file.foo.source = ./bar;` the store name would be decided by Nix, not HM.

**Comment by nonsequitur** [comment]
In this case we need to wrap the file with an extra derivation that outputs a symlink to the file.

```
if (source path contains home-manager marker and executable == null)
  link directly
else
  create wrapper
end
```
The first path is mostly taken, because most home-files have a source that was created through the `text` option and `executable` is null.

**Comment by nonsequitur** [comment]
To expand on that:

If we disallow setting `text` and `source` simultaneously (by removing `mkDefault` for `source`),
we have a very simple early exit in most cases:

```
if (text != null || (source path contains home-manager marker && executable == null))
   link directly
...
```

**Comment by rycee** [comment]
I think that is fair enough. But then we have the person who writes their own systemd service file and puts
```
xdg.configFile."systemd/user/myservice.service".source = ./mycoolservice.service;
```
in their configuration. If `~/.config/systemd/user/myservice.service` ends up being a link to the wrapped store path, which itself is a link then we're back to the original issue 🙂 (not that it works properly now).

**Comment by nonsequitur** [comment]
I don't see an elegant way to handle this, but to add another conditional case seems straightforward:
```
  ...
else if (source relative to home has prefix ".config/systemd/")
   copy source
else
  create wrapper
end
```
This can be later removed when systemd is fixed.

**Comment by rycee** [comment]
On the other hand, if systemd is special cased (which I don't particularly mind) then the current code could be made to cope with systemd using just

```patch
diff --git a/modules/files.nix b/modules/files.nix
index 42ef02d..b1b52b3 100644
--- a/modules/files.nix
+++ b/modules/files.nix
@@ -223,7 +223,8 @@ in
             install -m "$mode" "$source" "$target"
           else
             [[ -x $source ]] && isExecutable=1 || isExecutable=""
-            if [[ $executable == symlink || $isExecutable == $executable ]]; then
+            if [[ ($executable == symlink || $isExecutable == $executable)
+                && $relTarget != *"/systemd/user/"* ]]; then
               ln -s "$source" "$target"
             else
               cp "$source" "$target"
```

**Comment by nonsequitur** [comment]
Yes, but with the downside that we're copying all units for every generation.
I'd favor the single-file solution, but decision's on you. :blush:

**Comment by nonsequitur** [comment]
In an early draft, I used a magic `mtime` value to identify `home-file` symlinks (plus the fact that they point into the nix store), but I dismissed it as too brittle. Or do you think it could work reliably?

**Comment by rycee** [comment]
It might be a bit brittle, although I can't think of any particular reason at the moment. In any case I'm not particularly fond of the mtime idea in general since I'd like this information to be relatively explicit.

A possibility might be to do a pass over the old generation home-files and collect each file's store path in some searchable structure (that can be queried using a function like `isInOldGeneration` in the activation script). I.e., a list somewhat like produced by
```console
$ find /nix/var/nix/profiles/per-user/$USER/home-manager/home-files/ \! -type d -exec readlink -m '{}' ';' | sort -u
```

Then instead of
```bash
"$(readlink "$targetPath")" == ${homeFilePattern}
```
to check the target one could do
```bash
isInOldGeneration "$(readlink -m "$targetPath")"
```

I haven't thought this completely through, so there might be some fundamental flaw that I'm missing. But in principle I think it would work with a single link setup.

**Comment by nonsequitur** [comment]
Why not use [this way](https://github.com/rycee/home-manager/blob/master/modules/files.nix#L124) of checking if an old-gen file is present in the new-gen for the reverse check (`isInOldGeneration`)? Additionally, you can check if the destinations of the old and new links match.

The only advantage of a distinct home-file marker is extra robustness in the case that the old-gen home-files don't match up with the true state of the home-files in `$HOME`.
But because we don't make use of the markers to spot the most important cases of home-file corruption, we might as well get rid of them.<sup>1</sup>

A robust alternative to home-file markers could be an index file `~/.home-files` for all currently existing home-files.


<sup>1</sup>
For example, the following can happen:
Generation A is current and has no home files.
Generation B has file F.
We activate B, creating F in `$HOME`, but activation of B fails at the end.
A ist still current, so if we activate A again, F isn't removed.

**Comment by rycee** [comment]
Sorry for the delay! A while ago I added 7a9c8730931b0d6ec13ad115cd79826f30877d88 which improves the special case handling of systemd files. Basically, it now handles systemd files when given directly to `home.file.<name?>.source`, not just when using the systemd module. I think this resolves the original issue raised in this ticket.

About the cleanups you are right. I think the original reason why I used the pattern of the symlink target to determine whether a file was managed by HM was exactly that it was simple and quite robust with respect to "lost" managed files.

For the example scenario you gave I think it shouldn't be a problem because HM will switch over the current generation link before linking the files. So the following should hold

1. Generation A is current and has no home files.
2. Generation B has file F.
3. Activate script for B is run.
4. Just before linking the files in B we switch over the current home-manager profile in `/nix/var/nix/profiles` to B.
    1. If B fails now we can attempt to fix the problem and run B again and hope for the best.
5. Otherwise, F is linked to `$HOME`
    1. If B fails now then we can still attempt to fix the problem and run B again, we'll just overwrite the F link.
    2. We could also attempt to roll back to A, in which case F will be in the current generation and should be removed by A's activation script.

That said, the current implementation is having a problem if the old generation contains a file that the new generation does not have and the linking fails. Then the file will be left if we try activating the new generation again. I think this particular problem could be fixed by doing the cleanup step before switching over the current generation link.

Then we'd have

1. Generation A is current and has file F.
2. Generation B has no home file.
3. Activate script for B is run.
4. Cleanup of files in previous generation (the link to F is removed)
    1. If script fails now we can reactivate A which will recreate F or rerun B which will attempt to remove F again. Neither step should leave 
5. Switch over the current home-manager profile in `/nix/var/nix/profiles` to B.
    1. If script fails now we can reactivate A, which will again recreate F. Rerunning B will again attempt to remove F.

Essentially, if FA and FB are the set of files in generation A and B, respectively then what we do today when switching from A to B is FA → FA ∪ FB → (FA ∪ FB) ∩ FB = FB and the issue above arises when the A and B files are mixed together. If we instead clean before linking we'd have FA → FA ∩ FB → (FA ∩ FB) ∪ FB = FB, which should mean that we don't have this weird intermediate state with links in $HOME that belong to one generation but not the other.

I have not made this change yet, though. Will have to think it through a little bit more to make sure I didn't miss some unpleasant side effect

**Comment by nonsequitur** [comment]
Yeah, I came to the same conclusions about cleaning before linking, sorry for not adding this to my post right away.
Also, my example was unfitting, because the failure probability is negligibly small: Linking should almost always fail for all files or for none.

A more realistic case for home files to get out of sync with the generation is, for example, when the user restores an old backup of `$HOME`.  (But even then the matching home config to recreate the generation is probably stored in `$HOME`.)
Maybe we could use the `mtime` trick as an extra safety margin.

To sum up the design:
1. Link files directly
2. On activation:
  2.1. Clean
  2.2. Link generation to gcpath
  2.3. Link files
3. Maybe use `mtime` marker

**Comment by rycee** [comment]
The backup case is indeed interesting and I think you are right that it might be quite realistic. And as you mention I imagine that you'd typically restore to a home.nix version that matches the linked files, but if not then these links would not be visible to HM.

Further, when restoring a whole computer these links would also be pointing to some store path that no longer exists. The backup story might be worth noting in the README. I'll see if I can do that later.

I'll sort out the cleanup/link activation order but don't feel any particular pressing need for the direct linking. So I will leave that for you to do when you have the time ;-) Shouldn't be a problem getting it merged but I do want to highlight that I absolutely want to keep the `home-files` directory structure inside the generation directory and while I don't have anything particular against the mtime marker as a "fallback" solution I do want to keep some form of easy identification of symlinks that are managed by HM. For example, being able to do `readlink ~/.config/beets/config.yaml` and immediately seeing `home-manager-files` in the result is quite valuable to me, especially when trying to help people. I would not like to ask them to get the mtime and have to try to explain why it should be some specific odd value :-)

In any case, I think this particular issue with the systemd files has been resolved so I would suggest closing it and creating new issues/PRs for these other things.

**Comment by nonsequitur** [comment]
Agreed :relaxed:
